rust/kernel/sync/arc.rs

```diff
@@ -7,178 +7,218 @@
 //! underlying object when it reaches zero. It is also safe to use concurrently from multiple
 //! threads.
 //!
-//! It is different from the standard library's [`Arc`] in two ways: it does not support weak
-//! references, which allows it to be smaller -- a single pointer-sized integer; it allows users to
-//! safely increment the reference count from a single reference to the underlying object.
+//! It is different from the standard library's [`Arc`] in a few ways:
+//! 1. It is backed by the kernel's `refcount_t` type.
+//! 2. It does not support weak references, which allows it to be half the size.
+//! 3. It saturates the reference count instead of aborting when it goes over a threshold.
+//! 4. It does not provide a `get_mut` method, so the ref counted object is pinned.
 //!
 //! [`Arc`]: https://doc.rust-lang.org/std/sync/struct.Arc.html
 
-use crate::Result;
+use crate::{bindings, Result};
 use alloc::boxed::Box;
 use core::{
-    mem::ManuallyDrop,
-    ops::Deref,
+    cell::UnsafeCell, convert::AsRef, marker::PhantomData, mem::ManuallyDrop, ops::Deref, pin::Pin,
     ptr::NonNull,
-    sync::atomic::{fence, AtomicUsize, Ordering},
 };
 
+extern "C" {
+    fn rust_helper_refcount_new() -> bindings::refcount_t;
+    fn rust_helper_refcount_inc(r: *mut bindings::refcount_t);
+    fn rust_helper_refcount_dec_and_test(r: *mut bindings::refcount_t) -> bool;
+}
```

Comment:
Are these not exported from the `bindings` module?

Reply:
I don't think so, they're a `static inline` defined in a header unfortunately:
https://github.com/Rust-for-Linux/linux/blob/a5a22ed136b8b032675aff92de214a7032d1ddad/include/linux/refcount.h#L265-L268
https://github.com/Rust-for-Linux/linux/blob/a5a22ed136b8b032675aff92de214a7032d1ddad/include/linux/refcount.h#L331-L334

Reply:
I mean the `rust_helper_*` methods. Or are those invisible to bindgen?

Reply:
I think the `rust_helper_*` methods are a work-around to be able to call header inline/static functions. bindgen does not export them. So we create a C helper function in `helpers.c` which in turn calls the inline/static header function. Now we can call the helper function in Rust.

Reply:
It should work through bindgen too. We have #352 and #359.

Reply:
I'll move along with this as it's the current state of the art (however ugly it may be). We can update it once we have more automated support.